Source/Core/Core/HW/EXI_Channel.cpp

```diff
@@ -228,27 +229,28 @@ void CEXIChannel::DoState(PointerWrap &p)
 	p.Do(m_Control);
 	p.Do(m_ImmData);
 
-	for (int d = 0; d < NUM_DEVICES; ++d)
+	for (int device_index = 0; device_index < NUM_DEVICES; ++device_index)
 	{
-		IEXIDevice* pDevice = m_pDevices[d].get();
-		TEXIDevices type = pDevice->m_deviceType;
+		std::unique_ptr<IEXIDevice>& device = m_pDevices[device_index];
+		TEXIDevices type = device->m_deviceType;
 		p.Do(type);
-		IEXIDevice* pSaveDevice = (type == pDevice->m_deviceType) ? pDevice : EXIDevice_Create(type, m_ChannelId);
-		pSaveDevice->DoState(p);
-		if (pSaveDevice != pDevice)
+
+		if (type == device->m_deviceType)
 		{
-			// if we had to create a temporary device, discard it if we're not loading.
+			device->DoState(p);
+		}
+		else
+		{
+			// If we had to create a temporary device, discard it if we're not loading.
 			// also, if no movie is active, we'll assume the user wants to keep their current devices
 			// instead of the ones they had when the savestate was created,
 			// unless the device is NONE (since ChangeDevice sets that temporarily).
 			if (p.GetMode() != PointerWrap::MODE_READ)
-			{
-				delete pSaveDevice;
-			}
-			else
-			{
-				AddDevice(pSaveDevice, d, false);
-			}
+				return;
```

Comment:
This return statement is un-reachable. 

If PointerWrap is MODE_WRITE, then type can't change on line 236.